rfc/rfc-0009-controller-unit-testing.md

```diff
@@ -0,0 +1,84 @@
+# RFC-0009: Controller Unit Testing
+
+**Authors:** Glyn Normington
+
+**Status:**
+
+**Pull Request URL:** https://github.com/projectriff/riff/pull/1372
+
+**Superseded by:** N/A
+
+**Supersedes:** N/A
+
+**Related:** N/A
+
+
+## Problem
+The Kubernetes controllers in riff need unit testing in a consistent way. Other projects
+have adopted and promoted various ways of testing controllers. For ease of maintenance,
+riff needs to adopt a single, consistent approach.
+
+For the purpose of this document, unit testing refers to testing the
+`Reconcile` method of a controller.
+
+Apart from checking that various objects are created etc., it should be straigthtforward
+to test error paths by injecting errors in object lookup and creation.
+
+### Anti-Goals
+
+None.
+
+## Solution
+
+The approach taken is to use a fake client and other interfaces and introduce a test framework
+which allows tests to be written in a declarative style. See "Rejected Alternatives" below for 
+some other approaches which were tried out.
+
+The test framework supports table-driven tests. Each row of the table tests a single call to `Reconcile` by
+setting up some initial Kubernetes objects, modifying the client behaviour using various hooks
+or reactors, and then asserting that various Kubernetes objects are created, updated, etc.
+
+When a helper function (a utility function provided by a test) is used to construct a Kubernetes object, it should
+construct a fixed value and any variations should be explicitly coded in the table row. This avoids details of a
+particular test being "buried" in helper functions.
+
+Since kubebuilder was used to generate the controllers, some packages contain more than one controller and so
+helper functions need to be named or organised carefully to avoid clashes. 
+
+## Rejected Alternatives
```

Comment:
maybe add the approach with Gomega/Ginkgo to the list?

Reply:
Added. I think that helps to explain the rationale for the table-driven approach too.